src/Grassmann.jl

```diff
@@ -0,0 +1,275 @@
+Then the exponential map is written using
+````math
+z = x V\cos(S){\bar V}^\mathrm{T} + U\sin(S){\bar V}^\mathrm{T},
+````
+where ${\bar\cdot}^{\mathrm{T}}$ denotes the complex conjugate transposed.
+The cosine and sine are applied element wise to the diagonal entries of $S$.
+A final QR decomposition $z=QR$ is performed for numerical stability reasons,
+yielding the result as
+````math
+y = Q.
+````
+"""
+function exp!(M::Grassmann{N,K,T},y, x, v) where {N,K,T}
+    if norm(M,x,v) ≈ 0
+        return (y .= x)
+    end
+    d = svd(v)
+    z =  x * d.V * Diagonal(cos.(d.S)) * d.Vt + d.U * Diagonal(sin.(d.S)) * d.Vt
+    # reorthonormalize
+    y = copyto!(y, Array(qr(z).Q) )
+    return y
+end
+
+@doc doc"""
+    inner(M,x,v,w)
+
+compute the inner product for two tangent vectors `v`, `w` from the
+tangent space of `x` on the [`Grassmann`](@ref) manifold `M`.
+The formula reads
+````math
+(v,w)_x = \operatorname{trace}({\bar v}^{\mathrm{T}}w),
+````
+where ${\bar\cdot}^{\mathrm{T}}$ denotes the complex conjugate transposed.
+"""
+inner(::Grassmann{N,K,T}, x, v, w) where {N,K,T} = real(dot(v,w))
+
+@doc doc"""
+    inverse_retract!(M, v, x, y, ::PolarInverseRetraction)
+
+compute the inverse retraction for the [`PolarRetraction`](@ref)
+
+````math
+\operatorname{retr}_x^{-1}y = y*(\bar{x}^\mathrm{T}y)^{-1} - x,
+````
+where ${\bar\cdot}^{\mathrm{T}}$ denotes the complex conjugate transposed.
+"""
+inverse_retract!(::Grassmann{N,K,T}, v, x, y, ::PolarInverseRetraction) where {N,K,T} = ( v .= y/(x'*y) - x)
+
+@doc doc"""
+    inverse_retract!(M, v, x, y, ::QRInverseRetraction)
+
+compute the inverse retraction valid of the [`QRRetraction`](@ref)
+
+````math
+\operatorname{retr}_x^{-1}y = y*(\bar{x}^\mathrm{T}y)^{-1} - x,
+````
+where ${\bar\cdot}^{\mathrm{T}}$ denotes the complex conjugate transposed.
+"""
+inverse_retract!(::Grassmann{N,K,T}, v, x, y, ::QRInverseRetraction) where {N,K,T} = ( v .= y/(x'*y) - x)
+
+isapprox(M::Grassmann{N,K,T}, x, y; kwargs...) where {N,K,T} = isapprox(distance(M,x,y),0.; kwargs...)
+
+@doc doc"""
+    log!(M, v, x, y)
+
+compute the logarithmic map on the [`Grassmann`](@ref) manifold
+$\mathcal M=\mathrm{Gr}(n,k)$, i.e. the tangent vector `v` whose corresponding
+[`geodesic`](@ref) starting from `x` reaches `y` after time 1 on `M`. The formula reads
+````math
+v = V\cdot \operatorname{atan}(S) \cdot {\bar U}^\mathrm{T},
+````
+where ${\bar\cdot}^{\mathrm{T}}$ denotes the complex conjugate transposed,
+$U$ and $V$ are the unitary matrices, and $S$ is a diagonal matrix containing
+the singular values of the SVD-decomposition of
+````math
+USV = ({\bar y}^\mathrm{T}x)^{-1} ( {\bar y}^\mathrm{T} - {\bar y}^\mathrm{T}x{\bar x}^\mathrm{T} ).
+````
+In this formula the $\operatorname{atan}$ is meant elementwise.
+"""
+function log!(M::Grassmann{N,K,T}, v, x, y) where {N,K,T}
+    z = y'*x
+    At = y' - z*x'
+    Bt = z\At
+    d = svd(Bt')
+    return (v .= d.U * Diagonal(atan.(d.S)) * d.Vt)
```

Comment:
I think it's safer to return `v` on a separate line (this might return the RHS).